### PR TITLE
0.10.0: Update to work with Nightshade client + Refactoring

### LINKED ISSUE #5: NEARLIB: no explicit error if trying to sign transaction, but no key is found.

### LINKED ISSUE #12: bs58 should be smarter
```
    /**
     * Encode a buffer as string using bs58
     * @param {Buffer} buffer 
     * @example
     * KeyPair.encodeBufferInBs58(key.publicKey)
     */
    static encodeBufferInBs58(buffer) {
        const bytes = Buffer.from(buffer);
        const encodedValue = bs58.encode(bytes);
        return encodedValue;
    }
```
Stuff like this should be inside the bs58.encode() function instead of outside, bs58 should auto determine the type is Buffer or Uint8Array.  Maybe use a forked bs58 with this update?

Or get a base58 lib that supports uint8array buffers.

### LINKED ISSUE #18: Better error reporting
Hard to interpret error messages like:
> Error: [-32603] Internal error: Tx (00) not found

### LINKED ISSUE #27: Clean-up 'make function calls' test in integration.test.js
@vgrichina commented on [Fri Feb 15 2019](https://github.com/nearprotocol/nearcore/issues/579)

Test is currently whole wall of code with few unrelated tests jammed together:

```js
    test('make function calls', async () => {
        const args = {
            'name': 'trex'
        };
        const viewFunctionResult = await nearjs.callViewFunction(
            contractName,
            'hello', // this is the function defined in hello.wasm file that we are calling
            args);
        expect(viewFunctionResult).toEqual('hello trex');

        var setCallValue = await generateUniqueString('setCallPrefix');
        const setArgs = {
            'value': setCallValue
        };
        const scheduleResult = await nearjs.scheduleFunctionCall(
            0,
            aliceAccountName,
            contractName,
            'setValue', // this is the function defined in hello.wasm file that we are calling
            setArgs);
        expect(scheduleResult.hash).not.toBeFalsy();
        await nearjs.waitForTransactionResult(scheduleResult);
        const getValueResult = await nearjs.callViewFunction(
            contractName,
            'getValue', // this is the function defined in hello.wasm file that we are calling
            {});
        expect(getValueResult).toEqual(setCallValue);

        // test that load contract works and we can make calls afterwards
        const options = {
            viewMethods: ['hello', 'getValue'],
            changeMethods: ['setValue'],
            sender: aliceAccountName
        };
        const contract = await nearjs.loadContract(contractName, options);
        const helloResult = await contract.hello(args);
        expect(helloResult).toEqual('hello trex');
        var setCallValue2 = await generateUniqueString('setCallPrefix');
        const setArgs2 = {
            'value': setCallValue2
        };
        const setValueResult = await contract.setValue(setArgs2);
        expect(setValueResult.status).toBe('Completed');
        const getValueResult2 = await contract.getValue();
        expect(getValueResult2).toEqual(setCallValue2);
    });
```

compare to other tests, e.g.:

```js
    test('can get logs from method result', async () => {
        await contract.generateLogs();
        expect(logs).toEqual([`[${contractName}]: LOG: log1`, `[${contractName}]: LOG: log2`]);
    });
```

---

@ilblackdragon commented on [Tue Mar 26 2019](https://github.com/nearprotocol/nearcore/issues/579#issuecomment-476840403)

@janedegtiareva any update here?

### PR DESCRIPTION
**Pull request by ilblackdragon**
Refactors to use TypeScript, moves around some things to simplify usage of Account and Contract.

This follows ideas for #39 + TypeScript for better typed APIs.

Fixes #5
Fixes #12
Ref #13
Fixes #18
Fixes #27

### CODE REVIEW

File: test/integration.test.js [hunk 1]
```diff
@@ -38,110 +37,110 @@ devConnectDescribe('dev connect', () => {
...
-        window.localStorage = createFakeStorage();
+        window.localStorage = test_utils.createFakeStorage();
         // Mocking some
-        let tmpCreate = dev.createAccountWithContractHelper;
-        let devConfig = dev.getConfig;
-        dev.getConfig = async () => 'THE_CONFIG';
-        dev.createAccountWithContractHelper = async (nearConfig, newAccountId, publicKey) => {
+        let tmpCreate = test_utils.createAccountWithContractHelper;
+        let devConfig = test_utils.getConfig;
+        test_utils.getConfig = async () => 'THE_CONFIG';
```
Comment (vgrichina) [inline test/integration.test.js]: I think we actually don't need `getConfig` anymore, replaced by:
https://github.com/nearprotocol/near-shell/blob/master/blank_project/src/config.js

File: test/integration.test.js [hunk 2]
```diff
@@ -1,6 +1,5 @@
 const { Account, KeyPair, Near, InMemoryKeyStore } = require('../');
-const  { aliceAccountName, newAccountCodeHash, storageAccountIdKey, createFakeStorage, sleep } = require('./test-utils');
-const dev = require('../dev');
+const test_utils = require('./test-utils');
```
Comment (vgrichina) [inline test/integration.test.js]: s/test_utils/testUtils/g

File: test/test-utils.js [hunk 3]
```diff
@@ -27,4 +44,101 @@ function sleep(time) {
...
+    createAccountWithLocalNodeConnection: async function (newAccountName, newAccountPublicKey) {
+        const account = new Account(this.near.nearClient);
+        this.deps.keyStore.setKey(testAccountName, devKey); // need to have dev account in key store to use this.
+        const createAccountResponse = await account.createAccount(newAccountName, newAccountPublicKey, 1000000000000, testAccountName);
+        await this.near.waitForTransactionResult(createAccountResponse);
+    },
+    /**
+     * Function to create an account on near-hosted devnet using contract helper. This will not work on non-dev environments.
+     */
+    createAccountWithContractHelper: async function (newAccountId, publicKey) {
```
Comment (vgrichina) [inline test/test-utils.js]: this is needed outside of tests, e. g. for near-shell

**Comment by ilblackdragon** [comment]
The goal from now on to never have hard coded keys for regular usage (e.g. for now kept it for unit tests but ideally even unit tests would be parameterized with current key).

So let's figure out what the interface should look like: for example reading available configs for local nodes or asking user to provide it. But definitely now "devKey" → as this doesn't work in normal situation of me running actual testnet node locally.

Best regards,
Illia Polosukhin

Sent via Superhuman ( https://sprh.mn/?vip=ilblackdragon@gmail.com )

File: test/integration.test.js (hunk 2 above)
Comment (ilblackdragon) [reply test/integration.test.js]: Should the file also be named camel case style?

Comment (vgrichina) [reply test/integration.test.js]: @ilblackdragon I think dash is more common for module names.

**Pull request merged**